Added endpoint that, given series query parameters, returns ranked so…

sempervirescent opened the pull request:
…urces.

doriapenciucgro committed:
Added endpoint that, given series query parameters, returns ranked sources.

doriapenciucgro committed:
Added support to get ranked sources in batch.

nemozen commented on api/client/lib.py:
```diff
@@ -443,6 +443,24 @@ def get_data_series(access_token, api_host, **selection):
+# TODO: Add code to handle no query parameters call to v2/available/sources. It is hanging at the moment.
+
+
+def get_source_ranking(access_token, api_host, series):
```
this function returns the same thing as rank_series_by_source....

sempervirescent replied:
No, it does not. get_source_ranking is called inside rank_series_by_source. The first one calls the endpoint directly and returns an array of ranked sources. The latter an array of series with all information. I separated them so we can also get the ranking only directly, without the rest of the data.

nemozen committed:
in find_data_series(), yield all possible series  in order instead of just the highest ranked one

add_data_series() will still only use the first one, but this gives
more flexibility when using the lower level function find_data_series

sempervirescent commented:
@itsjohnward: clarified all comments by @nemozen. If you have a chance to take a look, would you let me know if you have any concerns? If not, could we merge? Thank you.

nemozen commented on api/client/__init__.py:
```diff
@@ -40,9 +40,11 @@ def lookup_belongs(self, entity_type, entity_id):
         return lib.lookup_belongs(self.access_token, self.api_host,
                                   entity_type, entity_id)
 
+    def get_source_ranking(self, series):
```
this method Client.get_source_ranking is not used ...

sempervirescent replied:
It is used in gro derived.

nemozen commented on api/client/__init__.py:
```diff
@@ -40,9 +40,11 @@ def lookup_belongs(self, entity_type, entity_id):
     def rank_series_by_source(self, series_list):
-        return lib.rank_series_by_source(self.access_token, self.api_host,
-                                         series_list)
+        return lib.rank_series_by_source(self.access_token, self.api_host, series_list)
```
lines should be < 80 chars as much as possible.

sempervirescent replied:
Reverted.

nemozen commented on api/client/batch_client.py:
```diff
@@ -27,6 +27,7 @@ def __init__(self, api_host, access_token):
         super(BatchClient, self).__init__(api_host, access_token)
         self._logger = lib.get_default_logger()
         self._http_client = AsyncHTTPClient()
+        self.path = ''
```
unused variable

sempervirescent replied:
Removed.

nemozen commented on api/client/batch_client.py:
```diff
@@ -80,20 +81,33 @@ def get_data(self, url, headers, params=None):
-        of: item_id, metric_id, region_id, frequency_id, source_id,
-        partner_region_id. Additional arguments are allowed and ignored.
+        """Get all the data points for a given selection, which is some or all of: item_id, metric_id, region_id,
+        frequency_id, source_id, partner_region_id. Additional arguments are allowed and ignored.
```
prefer to keep lines < 80  chars and also to avoid whitespace changes unless they are an improvement

sempervirescent replied:
Reverted.

nemozen commented on api/client/batch_client.py:
```diff
@@ -80,20 +81,33 @@ def get_data(self, url, headers, params=None):
+        response['ranked_sources'] = json_decode(resp)
+        raise gen.Return(response)
+
+    def batch_async_get_ranked_sources(self, batched_args, output_list=None, map_result=None):
```
series ranking does not benefit from batch client's async processing... so we don't need a different implementation. let's just make BatchCLient a subclass of GroClient and it will inherit this functionality

sempervirescent replied:
@nemozen we discussed this yesterday. I have both individual api calls tested in gro derived, as well as batch. Batch takes 28s as opposed to 1min 15sec on average for individual api calls. Are you suggesting that I should keep the individual api calls although they take a lot longer?

sempervirescent replied:
I am not seeing any functionality in GroClient that deals with ranking. What did you refer to when you mentioned that we can inherit this functionality from there? Thx

sempervirescent replied:
This was taken offline.

nemozen commented on api/client/lib.py:
```diff
@@ -443,6 +443,24 @@ def get_data_series(access_token, api_host, **selection):
+    return get_data(url, headers, params).json()
+
+
 def rank_series_by_source(access_token, api_host, series_list):
```
added support for time ranges in  https://github.com/gro-intelligence/api-client/pull/125 
and also separating finding frmo ranking..

sempervirescent replied:
@nemozen I don't understand this comment.

nemozen replied:
just fyi  that this function was changing to also support time range as part of the ranking, which it didn't before.  merged into this now.

nemozen committed:
make BatchCLient a subclass of GroClient

instead of subclassing the legacy Client class, this uses GroClient which has methods for discovery and series manipulation

nemozen committed:
revert unused method

nemozen committed:
remove unused variable and clean up whitespace

nemozen committed:
whitespace

nemozen committed:
Merge remote-tracking branch 'origin/development' into CLEWS-22341

nemozen committed:
define make_key before using it

nemozen committed:
remove spurious character

nemozen committed:
get_descendant_regions: absence of descendant_level is equivalent to "any"

this allows us to get all the levels

nemozen committed:
remove TODO

rank_series_by_source actually works ok with empty query selection, it just takes a long time.

for data_series in client.rank_series_by_source([{}]):
    print('data_series', data_series)

outputs:

('data_series', {'source_id': 2})
('data_series', {'source_id': 10})
('data_series', {'source_id': 35})
('data_series', {'source_id': 20})
('data_series', {'source_id': 43})
('data_series', {'source_id': 126})
('data_series', {'source_id': 15})
('data_series', {'source_id': 36})
('data_series', {'source_id': 59})
('data_series', {'source_id': 22})
('data_series', {'source_id': 3})
('data_series', {'source_id': 25})
('data_series', {'source_id': 1})
('data_series', {'source_id': 110})
('data_series', {'source_id': 114})
('data_series', {'source_id': 105})
('data_series', {'source_id': 54})
('data_series', {'source_id': 26})
('data_series', {'source_id': 8})
('data_series', {'source_id': 29})
('data_series', {'source_id': 39})
('data_series', {'source_id': 88})
('data_series', {'source_id': 56})
('data_series', {'source_id': 78})
('data_series', {'source_id': 28})
('data_series', {'source_id': 80})
('data_series', {'source_id': 33})
.....

nemozen committed:
Merge branch 'CLEWS-22341' of https://github.com/gro-intelligence/api-client into CLEWS-22341

nemozen committed:
in BatchClient get_ranked_sources use rank_series_by_source from parent class rather than duplicating it

nemozen committed:
remove unused import and clean up line width

nemozen committed:
rename *get_ranked_sources to *rank_series_by_source for consistency with non batch method

in general, makes it clearer to follow foo(), sync_foo() batch_async_foo()

nemozen committed:
Merge remote-tracking branch 'origin/development' into CLEWS-22341

nemozen reviewed (approved):
made a few changes and retested, looks good. the batch and and non-batch version  return the same results.

```
for data_series in client.rank_series_by_source(series_list):
    print(data_series)

for batch_resp in client.batch_async_rank_series_by_source(
        [{'series_list': series_list}]):
    for data_series in batch_resp:
        print(data_series)

```

sempervirescent commented on api/client/gro_client.py:
```diff
@@ -129,6 +129,7 @@ def find_data_series(self, **kwargs):
             search_results.append(
                 self.search('regions', kwargs['partner_region'])[:cfg.MAX_RESULT_COMBINATION_DEPTH])
             keys.append('partner_region_id')
+        all_data_series = []
```
@nemozen The changes in these files seem to come from another unrelated PR https://github.com/gro-intelligence/api-client/pull/127

sempervirescent commented on api/client/lib.py:
```diff
@@ -452,6 +452,25 @@ def get_data_series(access_token, api_host, **selection):
         raise Exception(resp.text)
 
 
+def get_source_ranking(access_token, api_host, series):
```
@nemozen Based on our conversation on chat, do we want to add back the reference to this that was removed from Client?
```
def get_source_ranking(self, series):
    return lib.get_source_ranking(self.access_token, self.api_host, series)
```

nemozen commented:
yes it's a prereq

On Tue, Nov 19, 2019 at 10:25 AM sempervirescent <notifications@github.com>
wrote:

> *@sempervirescent* commented on this pull request.
> ------------------------------
>
> In api/client/gro_client.py
> <https://github.com/gro-intelligence/api-client/pull/122#discussion_r347989912>
> :
>
> > @@ -129,6 +129,7 @@ def find_data_series(self, **kwargs):
>              search_results.append(
>                  self.search('regions', kwargs['partner_region'])[:cfg.MAX_RESULT_COMBINATION_DEPTH])
>              keys.append('partner_region_id')
> +        all_data_series = []
>
> @nemozen <https://github.com/nemozen> The changes in these files seem to
> come from another unrelated PR #127
> <https://github.com/gro-intelligence/api-client/pull/127>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gro-intelligence/api-client/pull/122?email_source=notifications&email_token=ADTN4VA3Z7MPSZEF3B62BYTQUQAO5A5CNFSM4JILZ6J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCMCW26Q#pullrequestreview-319122810>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADTN4VDAMMEDPHCFOX3LN73QUQAO5ANCNFSM4JILZ6JQ>
> .
>

nemozen commented:
no i think find_data_series does what you want..

On Tue, Nov 19, 2019 at 10:53 AM sempervirescent <notifications@github.com>
wrote:

> *@sempervirescent* commented on this pull request.
> ------------------------------
>
> In api/client/lib.py
> <https://github.com/gro-intelligence/api-client/pull/122#discussion_r348008616>
> :
>
> > @@ -452,6 +452,25 @@ def get_data_series(access_token, api_host, **selection):
>          raise Exception(resp.text)
>
>
> +def get_source_ranking(access_token, api_host, series):
>
> @nemozen <https://github.com/nemozen> Based on our conversation on chat,
> do we want to add back the reference in Client to this method?
> def get_source_ranking(self, series): return
> lib.get_source_ranking(self.access_token, self.api_host, series)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gro-intelligence/api-client/pull/122?email_source=notifications&email_token=ADTN4VBKRJEBVUAOBMASVVLQUQDXZA5CNFSM4JILZ6J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCMC4T6Q#pullrequestreview-319146490>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADTN4VHXHGSUMAVLSKHAS3DQUQDXZANCNFSM4JILZ6JQ>
> .
>
